RecoLocalTracker/SiPixelRecHits/plugins/SiPixelRecHitHeterogeneous.cc

```diff
@@ -179,15 +182,17 @@ void SiPixelRecHitHeterogeneous::acquireGPUCuda(const edm::HeterogeneousEvent& i
 
   edm::Handle<reco::BeamSpot> bsHandle;
   iEvent.getByToken( tBeamSpot, bsHandle);
-  float bs[3] = {0.f};
+  edm::Service<CUDAService> cs;
+  auto bs = cs->make_host_unique<float[]>(3, cudaStream);
```

Comment:
instead of allocating this at every event, can we allocate it once (per edm stream), with the `cudaHostAllocPortable` and `cudaHostAllocWriteCombined` flags ?

Reply:
It seems the CUDA documentation is a bit confusing, at least for what concerns memory allocations, access, etc.

[The documentation](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__HIGHLEVEL.html#group__CUDART__HIGHLEVEL_1gd5c991beb38e2b8419f50285707ae87e) for `cudaMallocHost` says:

>   - `cudaHostAllocPortable`: The memory returned by this call will be considered as pinned memory by all CUDA contexts, not just the one that performed the allocation.
>   - `cudaHostAllocWriteCombined`: Allocates the memory as write-combined (WC). WC memory can be transferred across the PCI Express bus more quickly on some system configurations, but cannot be read efficiently by most CPUs. WC memory is a good option for buffers that will be written by the CPU and read by the device via mapped pinned memory or host->device transfers.

However, assuming we use Unified Addressing, [section 5.11](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__UNIFIED.html) says:

> All host memory allocated through all devices using `cudaMallocHost()` and `cudaHostAlloc()` is always directly accessible from all devices that support unified addressing. This is the case regardless of whether or not the flags `cudaHostAllocPortable` and `cudaHostAllocMapped` are specified. 

but

> Note that this is not the case for memory allocated using the flag `cudaHostAllocWriteCombined`, as discussed below. 

and

> Not all memory may be accessed on devices through the same pointer value through which they are accessed on the host. These exceptions are host memory registered using `cudaHostRegister()` and host memory allocated using the flag `cudaHostAllocWriteCombined`. For these exceptions, there exists a distinct host and device address for the memory. The device address is guaranteed to not overlap any valid host pointer range and is guaranteed to have the same value across all devices that support unified addressing.

Reply:
However, it is not clear to me what is the usefulness of this *Unified Addressing* (not to be confused with the *Unified Memory* returned by `cudaMallocManaged()`).

Reply:
> So:
> 
> * `cudaHostAllocWriteCombined` should be useful for host-to-device memory copies, but needs to be benchmarked (again ?)
> * `cudaHostAllocPortable` is unnecessary ?

@fwyzard, I agree with your conclusion.

(and contrary to what I wrote in https://github.com/cms-patatrack/cmssw/pull/245#discussion_r247626279, already our current `cudaMallocHost()`-allocated pinned host memory can be used for transfers to any device, so that is not the effect of `cudaAllocPortable`)

Reply:
> `cudaHostAllocWriteCombined` should be useful for host-to-device memory copies, but needs to be benchmarked (again ?)

Using it only for beamspot likely makes no difference, so I'll change all our host-to-device copies to use that and we can benchmark. (I suspect the effect could be tiny as we don't transfer that much of data in a single transfer)